### ISSUE TITLE
SameDiff: Training evaluation on variables not required for loss doesn't work

### ISSUE THREAD

**Issue by eraly** [deeplearning4j/deeplearning4j#8861]
Errors out [here.](https://github.com/eclipse/deeplearning4j-examples/blob/ec9ab1b62c410c0d731d3de800ab845ec734cf8b/dl4j-examples/src/main/java/org/deeplearning4j/examples/samediff/training/SameDiffCustomListenerExample.java#L55)

Tried a bunch of things following the unit tests [here](https://github.com/eclipse/deeplearning4j/blob/2bd69c004c8bad625e02eaade653fd4095a38972/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/samediff/SameDiffTrainingTest.java#L52) and still couldn't get it to work. Looks like a bug...Needs further investigation.

**Comment by AlexDBlack** [comment]
Transferred to eclipse/deeplearning4j - confirming this is a bug still present on master.

**Comment by rnett** [comment]
I ran into something similar a while back, if it is the same thing and my notes from then are accurate, I think that the loss variable required by the listener was lost somewhere and wasn't passed to the inference session (because it only needs to calculate the gradient of the loss for training).

**Comment by AlexDBlack** [comment]
Turns out the example has a bug - combined with a minor bug in SameDiff - cc @eraly 
https://github.com/eclipse/deeplearning4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/samediff/training/SameDiffMNISTTrainingExample.java#L70-L78

Simply put: the softmax variable isn't required to calculate the loss value, hence it isn't calculated.
That's obviously incorrect - it should be:
```
SDVariable out = sd.nn().softmax("out", z, 1);
SDVariable loss = sd.loss().softmaxCrossEntropy("loss", label, out);
```

I've implemented a fix in SameDiff so that the training evaluation variables are always calculated, even if not needed for calculating the loss.
